app/models/cart_item.rb

```diff
@@ -28,6 +32,11 @@ def to_h
           total_with_vat: total_with_vat
         }
       end
+
+      def eql?(other)
+        other.instance_of?(self.class) && other.product.id == product.id
+      end
+      alias == eql?
```

Comment:
http://www.skorks.com/2009/09/ruby-equality-and-object-comparison/

first from google, probably you will find better about comparing objects. Most important read about `<=>(other)`